### PR TITLE
Copy annotations into chart CR with chart-operator prefix

### PR DESCRIPTION
**Pull request by tomahawk28**
During looking into https://github.com/giantswarm/giantswarm/issues/8082, I found out we do not compare annotations between desired and current status. 

We would copy annotations from app CR only with `chart-operator.giantswarm.io` prefix and compare the value as well.

### CODE REVIEW

**Comment by tomahawk28** [comment]
Tested on `gauss`

File: service/controller/app/v1/resource/chart/desired.go [hunk 1]
```diff
@@ -73,11 +74,11 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
...
 	annotations := map[string]string{}
 
-	// ForceHelmUpgrade has been set for this app CR so this needs to be passed
-	// on to the chart CR.
-	val, ok := input[annotation.ForceHelmUpgrade]
-	if ok {
-		annotations[annotation.ForceHelmUpgrade] = val
+	for k, val := range input {
+		// Copy all annotations which has a prefix with chart-operator.giantswarm.io.
+		if strings.HasPrefix(k, annotation.ChartOperatorPrefix) {
```
Comment (tomahawk28) [inline service/controller/app/v1/resource/chart/desired.go]: Better to copy all annotation related to `chart-operator`

File: pkg/annotation/annotation.go [hunk 2]
```diff
@@ -3,15 +3,14 @@
 package annotation
 
 const (
+	// ChartOperatorPrefix is the prefix of the annotation that controls the logic inside chart-operator
```
Comment (rossf7) [inline pkg/annotation/annotation.go]:
```suggestion
	// ChartOperatorPrefix is the prefix for annotations that control logic inside chart-operator.
```

File: service/controller/app/v1/resource/chart/desired.go [hunk 3]
```diff
@@ -73,11 +74,11 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
 func generateAnnotations(input map[string]string) map[string]string {
 	annotations := map[string]string{}
 
-	// ForceHelmUpgrade has been set for this app CR so this needs to be passed
-	// on to the chart CR.
-	val, ok := input[annotation.ForceHelmUpgrade]
-	if ok {
-		annotations[annotation.ForceHelmUpgrade] = val
+	for k, val := range input {
```
Comment (rossf7) [inline service/controller/app/v1/resource/chart/desired.go]:
```suggestion
	for k, v := range input {
```

Let's try and be consistent with var naming. This is shorter but I'd also be fine with `key` and `val`.

File: service/controller/app/v1/resource/chart/resource.go [hunk 4]
```diff
@@ -76,6 +79,19 @@ func equals(current, desired *v1alpha1.Chart) bool {
 	if !reflect.DeepEqual(current.Labels, desired.Labels) {
 		return false
 	}
+	for k, desiredValue := range desired.Annotations {
```
Comment (rossf7) [inline service/controller/app/v1/resource/chart/resource.go]:
```suggestion

	for k, desiredValue := range desired.Annotations {
```

File: service/controller/app/v1/resource/chart/resource.go [hunk 5]
```diff
@@ -76,6 +79,19 @@ func equals(current, desired *v1alpha1.Chart) bool {
...
+		if !strings.HasPrefix(k, annotation.ChartOperatorPrefix) {
+			continue
+		}
+
+		if currentValue, ok := current.Annotations[k]; !ok {
+			return false
+		} else {
```
Comment (rossf7) [inline service/controller/app/v1/resource/chart/resource.go]: We return above. So do we need the nested if here?

File: service/controller/app/v1/resource/chart/desired.go (hunk 3 above)
Comment (tomahawk28) [reply service/controller/app/v1/resource/chart/desired.go]: Agreed, shorther version I would like.

**Pull request merged**